File: cli/internal/service/config_service.go
package service

import (
	"os"
	"path/filepath"

	"gopkg.in/yaml.v3"
)

// TTYConfig holds TTY-related configuration
type TTYConfig struct {
	MaxParallelSessions int `yaml:"max_parallel_sessions"`
	TerminalCloseDelay  int `yaml:"terminal_close_delay"`
}

// VSCodeConfig holds VSCode extension configuration
type VSCodeConfig struct {
	SyncInterval      int  `yaml:"sync_interval"`
	WatchFeatureFiles bool `yaml:"watch_feature_files"`
}

// Config holds all Claritask configuration
type Config struct {
	TTY    TTYConfig    `yaml:"tty"`
	VSCode VSCodeConfig `yaml:"vscode"`
}

// DefaultConfig returns the default configuration
func DefaultConfig() *Config {
	return &Config{
		TTY: TTYConfig{
			MaxParallelSessions: 3,
			TerminalCloseDelay:  1,
		},
		VSCode: VSCodeConfig{
			SyncInterval:      1000,
			WatchFeatureFiles: true,
		},
	}
}

// LoadConfig loads configuration from .claritask/config.yaml
func LoadConfig() (*Config, error) {
	config := DefaultConfig()

	// Try to load from file
	configPath := filepath.Join(".claritask", "config.yaml")
	data, err := os.ReadFile(configPath)
	if err != nil {
		// File doesn't exist, return defaults
		return config, nil
	}

	// Parse YAML
	if err := yaml.Unmarshal(data, config); err != nil {
		return config, nil // Return defaults on parse error
	}

	// Validate and clamp values
	if config.TTY.MaxParallelSessions < 1 {
		config.TTY.MaxParallelSessions = 1
	}
	if config.TTY.MaxParallelSessions > 10 {
		config.TTY.MaxParallelSessions = 10
	}

	if config.VSCode.SyncInterval < 100 {
		config.VSCode.SyncInterval = 100
	}

	return config, nil
}

// LoadConfigFrom loads configuration from a specific directory
func LoadConfigFrom(dir string) (*Config, error) {
	config := DefaultConfig()

	configPath := filepath.Join(dir, ".claritask", "config.yaml")
	data, err := os.ReadFile(configPath)
	if err != nil {
		return config, nil
	}

	if err := yaml.Unmarshal(data, config); err != nil {
		return config, nil
	}

	// Validate
	if config.TTY.MaxParallelSessions < 1 {
		config.TTY.MaxParallelSessions = 1
	}
	if config.TTY.MaxParallelSessions > 10 {
		config.TTY.MaxParallelSessions = 10
	}

	return config, nil
}

File: vscode-extension/src/configService.ts
import * as fs from 'fs';
import * as path from 'path';
import * as yaml from 'yaml';

export interface TTYConfig {
  max_parallel_sessions: number;
  terminal_close_delay: number;
}

export interface VSCodeConfig {
  sync_interval: number;
  watch_feature_files: boolean;
}

export interface ClaritaskConfig {
  tty: TTYConfig;
  vscode: VSCodeConfig;
}

export const DEFAULT_CONFIG: ClaritaskConfig = {
  tty: {
    max_parallel_sessions: 3,
    terminal_close_delay: 1,
  },
  vscode: {
    sync_interval: 1000,
    watch_feature_files: true,
  },
};

/**
 * Load configuration from .claritask/config.yaml
 */
export function loadConfig(workspacePath: string): ClaritaskConfig {
  const configPath = path.join(workspacePath, '.claritask', 'config.yaml');

  try {
    const content = fs.readFileSync(configPath, 'utf8');
    const loaded = yaml.parse(content) as Partial<ClaritaskConfig>;

    // Merge with defaults
    const config: ClaritaskConfig = {
      tty: { ...DEFAULT_CONFIG.tty, ...loaded?.tty },
      vscode: { ...DEFAULT_CONFIG.vscode, ...loaded?.vscode },
    };

    // Validate and clamp values
    if (config.tty.max_parallel_sessions < 1) {
      config.tty.max_parallel_sessions = 1;
    }
    if (config.tty.max_parallel_sessions > 10) {
      config.tty.max_parallel_sessions = 10;
    }
    if (config.vscode.sync_interval < 100) {
      config.vscode.sync_interval = 100;
    }

    return config;
  } catch {
    return DEFAULT_CONFIG;
  }
}

/**
 * Watch for config file changes
 */
export function watchConfig(
  workspacePath: string,
  callback: (config: ClaritaskConfig) => void
): fs.FSWatcher | null {
  const configPath = path.join(workspacePath, '.claritask', 'config.yaml');

  try {
    return fs.watch(configPath, (eventType) => {
      if (eventType === 'change') {
        const newConfig = loadConfig(workspacePath);
        callback(newConfig);
      }
    });
  } catch {
    return null;
  }
}
